src/visitpy/common/visitmodule.C

```diff
@@ -17264,19 +17282,50 @@ std::vector<PyMethodDef> VisItMethods;
 //   Cyrus Harrison, Wed Mar 17 11:16:58 PDT 2010
 //   Use PyCFunction typedef.
 //
+//   Cyrus Harrison, Mon Dec 14 11:28:57 PST 2020
+//   Pre-allocate the std::vector on first call to add.
+//   Since we hand pointers of these structs to python, we need
+//   to make sure the memory backing them isn't re/deallocated as we call
+//   push_back. This was not an issue with python 2 b/c how the module
+//   was inited, but realloc did cause problems with python 3 due 
+//   to required multi-step init.
+//
 // ****************************************************************************
 
 static void
 AddMethod(const char *methodName,
           PyCFunction methodImp,
           const char *doc = NULL)
 {
-    PyMethodDef newMethod;
-    newMethod.ml_name  = (char *)methodName;
-    newMethod.ml_meth  = methodImp;
-    newMethod.ml_flags = METH_VARARGS;
-    newMethod.ml_doc   = (char *)doc;
-    VisItMethods.push_back(newMethod);
+    // these structs are passed to python, which uses the pointers
+    // they provide.
+    // the memory backing them can't change!
+
+    if(VisItMethods.size() == 0)
+    {
+        VisItMethods.reserve(VISIT_METHODS_MAX_SIZE);
+    }
+
+    // make sure we don't realloc, error if we hit max
+    if(VisItMethods.size() < VISIT_METHODS_MAX_SIZE)
+    {
+        PyMethodDef newMethod;
+        newMethod.ml_name  = (char *)methodName;
+        newMethod.ml_meth  = methodImp;
+        newMethod.ml_flags = METH_VARARGS;
+        newMethod.ml_doc   = (char *)doc;
+        VisItMethods.push_back(newMethod);
+    }
+    else
+    {
+        // ERROR!
+        std::ostringstream emsg;
```

Comment:
I would rather see...

```
assert(VisItMethods.size() < VISIT_METHODS_MAX_SIZE);
PyMethodDef newMethod;
newMethod.ml_name  = (char *)methodName;
newMethod.ml_meth  = methodImp;
newMethod.ml_flags = METH_VARARGS;
newMethod.ml_doc   = (char *)doc;
VisItMethods.push_back(newMethod);
```

Why? Because this is an error *only* developers can fix. Asserts are compiled out with `-DCMAKE_BUILD_TYPE:STRING="Release`

What value does reporting a run-time error have when there is *nothing* the user can do about it? Also, isn't the answer to this question *staticly* determined for any VisIt release. I mean if the CLI inits properly for *any* user, then it will always init properly for every user? Or, can a user do something that is not part of our release that would cause the real number of methods to change?

Reply:
Apologies to @cyrush for my tone here. I just re-read and it reads more negative on the page than I felt in my head when I was writing it 😮 

My basic question is, does the additional exception and messaging logic really buy very much? Is the user likely to even see it? Can the conition be met for us in routine testing, etc. but wind up being violated for a user? I mean, is that even possible?

Reply:
If the condition happens, things are in a very bad state. The CLI will be in a pseudo broken state and honestly shouldn't be used.  It would be extremely hard to know what is wrong from a user report that doesn't point to it directly. 

With the current logic, we would never get a test suite pass if the assumptions are violated. 

Not sure the asserts provide that strong protection?

Another point:  Plugins do add methods. If someone has their own plugins, they could trigger regardless if our release is fine.

Reply:
> Another point: Plugins do add methods. If someone has their own plugins, they could trigger regardless if our release is fine.

Ok, can't use `assert()` then. Any way to do in two-passes...first to count, then to allocate and init?

Also, might wanna provide rationale for `512`.

BTW, I just counted `develop`...

```
% echo "dir()" | ./bin/visit -cli -nowin -forceinteractivecli | tr ',' '\n' | tr -d " '" | wc -l
Running: cli -dv -nowin -forceinteractivecli
Running: viewer -dv -nowin -noint -host 127.0.0.1 -port 5600
>>> >>> 
     450
```
If a user is running a few extra plugins, can they push over that `512` easily?

Reply:
@markcmiller86 no worries: as to if the user will see it. They will, the cli fails to start and you see the error message.

![Screen Shot 2020-12-14 at 2 32 21 PM](https://user-images.githubusercontent.com/1194526/102144183-e6aaf480-3e19-11eb-92e5-82a2021b65f4.png)

Reply:
I just re-ran my `develop`, enabled all plugins and now get `463`. Based on that, `512` feels a bit low.

Reply:
I can up it -- when I ran my build I saw 350 entries into the method table. Things in dir() may not 100% equal to things in the method table, but I don't see any reason not to up it -- It's only 32bytes per entry. 

1024?

Reply:
> 1024?

Sounds good.